From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Fri, 18 Feb 2022 10:02:01 +0100
Subject: Add dest_lookup and src_lookup

This allows to read and set the multicast destination and source lookup
settings on a bridge port in the kernel. This is used for PON.

Set it like this:
  $ bridge link set dev eth0_0 src_lookup on
  $ bridge link set dev eth0_0 dest_lookup off

look at the setting like this:
  $ bridge -d link show dev eth0_0
  5: eth0_0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9582 master sw257 state forwarding priority 32 cost 2 learning_limit -1
      hairpin off guard off root_block off fastleave off learning on flood on mcast_flood on isolated off dest_lookup off src_lookup on eth0_0
---
 bridge/link.c                | 28 +++++++++++++++++++++++++++-
 include/uapi/linux/if_link.h |  2 ++
 2 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/bridge/link.c b/bridge/link.c
index 29bab088..877e85ea 100644
--- a/bridge/link.c
+++ b/bridge/link.c
@@ -180,6 +180,14 @@ static void print_protinfo(FILE *fp, struct rtattr *attr)
 		if (prtb[IFLA_BRPORT_ISOLATED])
 			print_on_off(PRINT_ANY, "isolated", "isolated %s ",
 				     rta_getattr_u8(prtb[IFLA_BRPORT_ISOLATED]));
+
+		if (prtb[IFLA_BRPORT_DEST_LOOKUP])
+			print_on_off(PRINT_ANY, "dest_lookup", "dest_lookup %s ",
+				    rta_getattr_u8(prtb[IFLA_BRPORT_DEST_LOOKUP]));
+
+		if (prtb[IFLA_BRPORT_SRC_LOOKUP])
+			print_on_off(PRINT_ANY, "src_lookup", "src_lookup %s ",
+				    rta_getattr_u8(prtb[IFLA_BRPORT_SRC_LOOKUP]));
 	} else
 		print_stp_state(rta_getattr_u8(attr));
 }
@@ -281,6 +289,8 @@ static void usage(void)
 		"                               [ neigh_suppress {on | off} ]\n"
 		"                               [ vlan_tunnel {on | off} ]\n"
 		"                               [ isolated {on | off} ]\n"
+		"                               [ dest_lookup {on | off} ]\n"
+		"                               [ src_lookup {on | off} ]\n"
 		"                               [ hwmode {vepa | veb} ]\n"
 		"                               [ backup_port DEVICE ] [ nobackup_port ]\n"
 		"                               [ self ] [ master ]\n"
@@ -310,6 +320,8 @@ static int brlink_modify(int argc, char **argv)
 	__s8 mcast_flood = -1;
 	__s8 mcast_to_unicast = -1;
 	__s8 isolated = -1;
+	__s8 dest_lookup = -1;
+	__s8 src_lookup = -1;
 	__s8 hairpin = -1;
 	__s8 bpdu_guard = -1;
 	__s8 fast_leave = -1;
@@ -425,6 +437,17 @@ static int brlink_modify(int argc, char **argv)
 			isolated = parse_on_off("isolated", *argv, &ret);
 			if (ret)
 				return ret;
+		} else if (strcmp(*argv, "dest_lookup") == 0) {
+			NEXT_ARG();
+			dest_lookup = parse_on_off("dest_lookup", *argv, &ret);
+			if (ret)
+				return ret;
+
+		} else if (strcmp(*argv, "src_lookup") == 0) {
+			NEXT_ARG();
+			src_lookup = parse_on_off("src_lookup", *argv, &ret);
+			if (ret)
+				return ret;
 		} else if (strcmp(*argv, "backup_port") == 0) {
 			NEXT_ARG();
 			backup_port_idx = ll_name_to_index(*argv);
@@ -502,10 +525,13 @@ static int brlink_modify(int argc, char **argv)
 			 vlan_tunnel);
 	if (isolated != -1)
 		addattr8(&req.n, sizeof(req), IFLA_BRPORT_ISOLATED, isolated);
-
 	if (backup_port_idx != -1)
 		addattr32(&req.n, sizeof(req), IFLA_BRPORT_BACKUP_PORT,
 			  backup_port_idx);
+	if (dest_lookup != -1)
+		addattr8(&req.n, sizeof(req), IFLA_BRPORT_DEST_LOOKUP, dest_lookup);
+	if (src_lookup != -1)
+		addattr8(&req.n, sizeof(req), IFLA_BRPORT_SRC_LOOKUP, src_lookup);
 
 	addattr_nest_end(&req.n, nest);
 
diff --git a/include/uapi/linux/if_link.h b/include/uapi/linux/if_link.h
index 3dc6d1e7..744dd877 100644
--- a/include/uapi/linux/if_link.h
+++ b/include/uapi/linux/if_link.h
@@ -535,6 +535,8 @@ enum {
 	IFLA_BRPORT_MCAST_EHT_HOSTS_LIMIT,
 	IFLA_BRPORT_MCAST_EHT_HOSTS_CNT,
 	IFLA_BRPORT_LEARNING_LIMIT,
+	IFLA_BRPORT_DEST_LOOKUP,
+	IFLA_BRPORT_SRC_LOOKUP,
 	__IFLA_BRPORT_MAX
 };
 #define IFLA_BRPORT_MAX (__IFLA_BRPORT_MAX - 1)
